Dynamically decide which sdk to test against

gfontenot opened the pull request:
If we're testing iOS apps, we want to specify the simulator, which means we
need to know if we're trying to test an iOS app or an OS X app.

In order to do so, we can just look at the output of
`xcodebuild -showBuildSettings` for the chosen project/scheme. If we see a
setting for `CORRESPONDING_SIMULATOR_SDK_NAME`, then we know it's an iOS
project, and can target the `iphonesimulator` sdk. To be safe, if we
_don't_ find that setting, we're pointing at the `macosx` sdk explicitly.

Fixes #9

gfontenot committed:
Simplify calling bin scripts

 - `s:bin_script` simplifies building paths to the plugin's `bin`
   directory.
 - `s:quoted` quotes a provided string
 - `s:cli_arg` adds a space and uses `s:quoted` to wrap the provided
   string in quotes.

Applying these funcions to the scheme finding command.

gfontenot committed:
Dynamically decide which sdk to test against

If we're testing iOS apps, we want to specify the simulator, which means
we need to know if we're trying to test an iOS app or an OS X app.

In order to do so, we can just look at the output of
`xcodebuild -showBuildSettings` for the chosen project/scheme. If we see
a setting for `CORRESPONDING_SIMULATOR_SDK_NAME`, then we know it's an
iOS project, and can target the `iphonesimulator` sdk. To be safe, if we
_don't_ find that setting, we're pointing at the `macosx` sdk
explicitly.

gfontenot committed:
Use variadic arguments for cli args

Instead of needing to concatenate these manually, it makes more sense to
make this function take a variadic list of arguments and then quote them
automatically.

keith commented on plugin/xcodebuild.vim:
```diff
@@ -35,11 +53,28 @@ function! s:project_file()
+
+function! s:sdk()
+  if !exists('s:use_simulator')
+    let _ = system('source ' . s:bin_script('use_simulator.sh') . s:cli_args(s:project_file(), s:scheme_name()))
```
Is the `let _ =` necessary here?

gfontenot replied:
Surprisingly, yes. I'm not sure what's going on, but without it the sdk doesn't get set correctly. Maybe @christoomey or @gabebw knows why?

gabebw replied:
What if you replace `let _ = system(...)` with `call system(...)`?

In variable assignment, `call` isn't necessary to call functions, but when calling them by themselves, it is. Maybe that's what's happening here?

gfontenot replied:
Ah, yep, that fixes the error. Good to know, makes sense in retrospect.

keith commented on bin/use_simulator.sh:
```diff
@@ -0,0 +1,10 @@
+
+xcodebuild -project "$project" -scheme "$scheme" -showBuildSettings \
+  | grep CORRESPONDING_SIMULATOR_SDK_NAME \
+  2>&1 > /dev/null
```
Since this is after the pipe will stderr from xcodebuild still come through?

gfontenot replied:
As in: redirecting `stderr` is redundant here since the pipe will fail? I'm not sure how this should be written, honestly.

keith replied:
I was thinking that the redirect may only apply to grep's stderr

keith replied:
If that were try you'd need to do something like:

```
xcodebuild 2>/dev/null | grep 2>&1 >/dev/null
```

gfontenot replied:
yep, you're right. Good call.

gabebw replied:
For one, `2>&1 >/dev/null` can be `&>/dev/null`.

I think this can be cleaned up -- what exactly are you trying to do?

gfontenot replied:
I'm essentially trying to ignore any of the output. All I care about is the `grep` return status. Was there or wasn't there a key called `CORRESPONDING_SIMULATOR_SDK_NAME`.

gabebw replied:
> All I care about is the grep return status

Then I believe you want grep's `-q` option:

>  Quiet mode: suppress normal output.  grep will only search a file until a match has been found, making searches potentially less expensive.

```
xcodebuild -project "$project" -scheme "$scheme" -showBuildSettings 2>/dev/null \
  | grep -q CORRESPONDING_SIMULATOR_SDK_NAME
```

gabebw commented on bin/use_simulator.sh:
```diff
@@ -0,0 +1,10 @@
+
+set -o pipefail
+
+project="$1"
```
Interestingly, quotes aren't needed when setting variables.

gabebw commented on plugin/xcodebuild.vim:
```diff
@@ -3,13 +3,31 @@ command! XTest call <sid>test()
+endfunction
+
+function! s:quoted(string)
+  return '"' . a:string . '"'
```
I think you may want `shellescape`: `return shellescape(a:string)`

gfontenot replied:
oh dang. Yes, that's exactly what I wanted. Thanks!

gfontenot committed:
Use shellescape over out quoted function

shellescape is built in, and is actually what we wanted. We should rely
on that instead of our own custom solution.

gfontenot committed:
Use `call` instead of `let _ =`

I needed to use `let _ =` here because this script wasn't getting
executed properly without it. My lack of vimscript knowledge was
showing.

As @gabebw put it:

> In variable assignment, call isn't necessary to call functions, but
> when calling them by themselves, it is. Maybe that's what's happening
> here?

Changing this to `call` means the script is executed properly without
the weirdo hack.

gfontenot committed:
Simplify use_simulator script

As @keith pointed out, this wasn't properly redirecting errors from
xcodebuild, because of the pipe.

Then @gabebw came along, and suggested a much simpler solution, using
`grep`'s `-q` flag:

> Quiet mode: suppress normal output. grep will only search a file until
> a match has been found, making searches potentially less expensive.

I also removed the temp variable assignment and the pipefail, because I
didn't think they were super useful in this case.

gfontenot commented:
Thanks for the notes, guys!

gabebw commented on plugin/xcodebuild.vim:
```diff
@@ -3,13 +3,27 @@ command! XTest call <sid>test()
+function! s:cli_args(...)
+  let cli_args = ''
+
+  for cli_arg in a:000
```
So, this is kind of crazy, but I think you can replace this whole function with this per `:help map()`:

``` vim
let cli_args = join(map(a:000, shellescape(v:val)))
```

gabebw replied:
Er,

``` vim
return join(map(a:000, shellescape(v:val)))
```

gfontenot replied:
you telling me I should be using functors, gabe?

also, does `join` add a preceding space? I'm assuming not. So this would _probably_ have to be

``` vim
return ' ' . join(map(a:000, shellescape(v:val)))
```

right?

gabebw replied:
Yup, what you have looks correct.

gfontenot replied:
Hmm. This doesn't seem to want to work like this. `v:val` isn't defined. :confused:

gfontenot replied:
Oh, wild. You actually need to pass the function argument as a string. vimscript, you crazy.

gabebw replied:
Plus, I love that both `map()` and `map` exist and are TOTALLY different.
